test/xbgpu/test_engine.py

```diff
@@ -984,6 +1053,22 @@ def heap_factory(batch_index: int, present: np.ndarray) -> list[spead2.send.Heap
         delays[..., 0] = rng.uniform(-1e-9, 1e-9, size=(len(beam_outputs), n_ants))
         # Phase is in radians
         delays[..., 1] = rng.uniform(-2 * np.pi, 2 * np.pi, size=(len(beam_outputs), n_ants))
+
+        katcp_requests: dict[str, list[tuple[str, ...]]] = {
+            output.name: [
+                ("beam-weights", output.name, *weights[i]),
+                ("beam-quant-gains", output.name, quant_gains[i]),
+                ("beam-delays", output.name, *[f"{d[0]}:{d[1]}" for d in delays[i]]),
+            ]
+            for i, output in enumerate(beam_outputs)
+        }
+        flattened_requests = list(chain.from_iterable(katcp_requests.values()))
```

Comment:
Since you're only using the values of `katcp_requests` (and not the keys), why not make it a list of lists rather than a dict of lists?